src/main/java/com/openlattice/postgres/streams/BasePostgresIterable.kt

```diff
@@ -0,0 +1,191 @@
+package com.openlattice.postgres.streams
+
+import com.dataloom.streams.StreamUtil
+import com.google.common.base.Preconditions.checkState
+import com.zaxxer.hikari.HikariDataSource
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.io.Closeable
+import java.io.IOException
+import java.sql.*
+import java.util.NoSuchElementException
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Executors
+import java.util.concurrent.locks.ReentrantLock
+import java.util.function.Supplier
+import java.util.stream.Stream
+
+/**
+ *
+ * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
+ */
+class BasePostgresIterable<T>(
+        private val rsh: Supplier<StatementHolder>,
+        private val mapper: java.util.function.Function<ResultSet, T>
+) : Iterable<T> {
+
+    private val logger = LoggerFactory.getLogger(PostgresIterable::class.java)
+
+    override fun iterator(): PostgresIterator<T> {
+        try {
+            return PostgresIterator(rsh.get(), mapper)
+        } catch (e: SQLException) {
+            logger.error("Error creating postgres stream iterator.")
+            throw IllegalStateException("Unable to instantiate postgres iterator.", e)
+        } catch (e: IOException) {
+            logger.error("Error creating postgres stream iterator.")
+            throw IllegalStateException("Unable to instantiate postgres iterator.", e)
+        }
+
+    }
+
+    fun stream(): Stream<T> {
+        return StreamUtil.stream(this)
+    }
+
+
+}
+
+abstract class StatementHolderSupplier(
+        val hds: HikariDataSource, val sql: String,
+        val fetchSize: Int = 0,
+        val autoCommit: Boolean = fetchSize > 0
+) : Supplier<StatementHolder> {
+    init {
+        check(fetchSize >= 0) { "Fetch-size must be nonnegative." }
+        check(((!autoCommit) && (fetchSize > 0)) || autoCommit) {
+            "Auto-commit cannot be disabled if fetch size > 0."
+        }
+    }
+
+    protected val logger = LoggerFactory.getLogger(javaClass)
+
+    open fun execute(statement: Statement): ResultSet {
+        return statement.executeQuery(sql)
+    }
+
+    open fun buildStatement(connection: Connection): Statement {
+        return connection.createStatement()
+    }
+
+    override fun get(): StatementHolder {
+        val connection = hds.connection //okay to fail + propagate
+        val oldAutoCommit = connection.autoCommit
+        connection.autoCommit = autoCommit
+
+        val statement = buildStatement(connection) //okay to fail + propagate
+
+        statement.fetchSize = fetchSize
+
+        val rs = try {
+            execute(statement)
+        } catch (ex: Exception) {
+            logger.error("Error while executing sql: {}. The following exception was thrown: ", sql, ex)
+            connection.rollback()
+            throw ex
+        } finally {
+            connection.autoCommit = oldAutoCommit
```

Comment:
Not sure I understand why this is done since the connection object should be short-lived/not re-used (right?)

Reply:
I just do it for hygiene. Hikari will automatically reset this to pool default when it reclaims the connection, but if we ever switch connection pools or they ever change this behavior things won't explode. 

Plus it avoids the connection pool logging an unnecessary warning that it is resetting the autoCommit setting for the connection.

Reply:
Okay awesome! Makes perfect sense